README.md
CRUDlex Amazon S3 FileProcessor
===============================

This is a FileProcessor for [CRUDlex](https://github.com/philiplb/CRUDlex)
handling the uploaded files via Amazon S3 in CRUDlex versions not using
Flysystem for the file handling: >= 0.10.0 and <= 0.11.0.

## How To: Amazon S3 FileProcessor

First, create an instance of the Amazon S3 FileProcessor:

```php
$fileProcessor = new CRUDlex\AmazonS3FileProcessor(
    'yourRegion',
    'yourBucket',
    'yourAccessKey',
    'yourSecretAccessKey'
);
```

And then hand it in when registering the CRUDlex ServiceProvider:

```php
$app->register(new CRUDlex\ServiceProvider(), array(
    'crud.file' => __DIR__ . '<yourCrud.yml>',
    'crud.datafactory' => $dataFactory,
    'crud.fileprocessor' => $fileProcessor
));
```

## Package

[![Total Downloads](https://poser.pugx.org/philiplb/crudlexamazons3fileprocessor/downloads.svg)](https://packagist.org/packages/philiplb/crudlexamazons3fileprocessor)
[![Latest Stable Version](https://poser.pugx.org/philiplb/crudlexamazons3fileprocessor/v/stable.svg)](https://packagist.org/packages/philiplb/crudlexamazons3fileprocessor)
[![Latest Unstable Version](https://poser.pugx.org/philiplb/crudlexamazons3fileprocessor/v/unstable.svg)](https://packagist.org/packages/philiplb/crudlexamazons3fileprocessor) [![License](https://poser.pugx.org/philiplb/crudlexamazons3fileprocessor/license.svg)](https://packagist.org/packages/philiplb/crudlexamazons3fileprocessor)

### Stable

```json
"require": {
    "philiplb/crudlexamazons3fileprocessor": "0.11.0"
}
```

### Bleeding Edge

```json
"require": {
    "philiplb/crudlexamazons3fileprocessor": "0.12.x-dev"
}
```


CHANGELOG.md
CRUDlex Amazon S3 FileProcessor Changelog
=========================================

# 1.0.0
Released: 2017-06-29
- As CRUDlex uses Flysystem now, the maximum CRUDlex version is now 0.11.0

# 0.12.0
Released: 2017-03-26
- Relaxed a bit the required CRUDlex version

## 0.11.0
Released: 2016-09-21
- Updated the aws/aws-sdk-php to ~3.19 and so requiring the region as first parameter in the constructor
- Relaxed a bit the required CRUDlex version

## 0.10.0
Released: 2016-09-18
- Adjusted to the new CRUDlex 0.10.0 API
- Using the CRUDlex mimetype detection instead of an own
- Added inline documentation
- Switched to the array shorthand
- Attention: The minimum PHP version is now 5.5
- Attention: Switched from PSR-0 to PSR-4
- Added a Cache-Control header, thx JMF

## 0.9.10
Released: 2016-07-19
- Adjusted to the new CRUDlex 0.9.10 API

## 0.9.9
Released: 01.03.2016

First release.


src/CRUDlex/CRUDAmazonS3FileProcessor.php
<?php

/*
 * This file is part of the CRUDlex package.
 *
 * (c) Philip Lehmann-Böhm <philip@philiplb.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace CRUDlex;

use CRUDlex\CRUDFileProcessorInterface;
use CRUDlex\CRUDEntity;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpFoundation\StreamedResponse;
use Aws\S3\S3Client;

class CRUDAmazonS3FileProcessor implements CRUDFileProcessorInterface {

    protected $client;

    protected $bucket;

    protected function getMimeType($file) {

        // THX to
        // http://stackoverflow.com/questions/134833/how-do-i-find-the-mime-type-of-a-file-with-php

        // MIME types array
        $mimeTypes = array(
            "323"       => "text/h323",
            "acx"       => "application/internet-property-stream",
            "ai"        => "application/postscript",
            "aif"       => "audio/x-aiff",
            "aifc"      => "audio/x-aiff",
            "aiff"      => "audio/x-aiff",
            "asf"       => "video/x-ms-asf",
            "asr"       => "video/x-ms-asf",
            "asx"       => "video/x-ms-asf",
            "au"        => "audio/basic",
            "avi"       => "video/x-msvideo",
            "axs"       => "application/olescript",
            "bas"       => "text/plain",
            "bcpio"     => "application/x-bcpio",
            "bin"       => "application/octet-stream",
            "bmp"       => "image/bmp",
            "c"         => "text/plain",
            "cat"       => "application/vnd.ms-pkiseccat",
            "cdf"       => "application/x-cdf",
            "cer"       => "application/x-x509-ca-cert",
            "class"     => "application/octet-stream",
            "clp"       => "application/x-msclip",
            "cmx"       => "image/x-cmx",
            "cod"       => "image/cis-cod",
            "cpio"      => "application/x-cpio",
            "crd"       => "application/x-mscardfile",
            "crl"       => "application/pkix-crl",
            "crt"       => "application/x-x509-ca-cert",
            "csh"       => "application/x-csh",
            "css"       => "text/css",
            "dcr"       => "application/x-director",
            "der"       => "application/x-x509-ca-cert",
            "dir"       => "application/x-director",
            "dll"       => "application/x-msdownload",
            "dms"       => "application/octet-stream",
            "doc"       => "application/msword",
            "dot"       => "application/msword",
            "dvi"       => "application/x-dvi",
            "dxr"       => "application/x-director",
            "eps"       => "application/postscript",
            "etx"       => "text/x-setext",
            "evy"       => "application/envoy",
            "exe"       => "application/octet-stream",
            "fif"       => "application/fractals",
            "flr"       => "x-world/x-vrml",
            "gif"       => "image/gif",
            "gtar"      => "application/x-gtar",
            "gz"        => "application/x-gzip",
            "h"         => "text/plain",
            "hdf"       => "application/x-hdf",
            "hlp"       => "application/winhlp",
            "hqx"       => "application/mac-binhex40",
            "hta"       => "application/hta",
            "htc"       => "text/x-component",
            "htm"       => "text/html",
            "html"      => "text/html",
            "htt"       => "text/webviewhtml",
            "ico"       => "image/x-icon",
            "ief"       => "image/ief",
            "iii"       => "application/x-iphone",
            "ins"       => "application/x-internet-signup",
            "isp"       => "application/x-internet-signup",
            "jfif"      => "image/pipeg",
            "jpe"       => "image/jpeg",
            "jpeg"      => "image/jpeg",
            "jpg"       => "image/jpeg",
            "js"        => "application/x-javascript",
            "latex"     => "application/x-latex",
            "lha"       => "application/octet-stream",
            "lsf"       => "video/x-la-asf",
            "lsx"       => "video/x-la-asf",
            "lzh"       => "application/octet-stream",
            "m13"       => "application/x-msmediaview",
            "m14"       => "application/x-msmediaview",
            "m3u"       => "audio/x-mpegurl",
            "man"       => "application/x-troff-man",
            "mdb"       => "application/x-msaccess",
            "me"        => "application/x-troff-me",
            "mht"       => "message/rfc822",
            "mhtml"     => "message/rfc822",
            "mid"       => "audio/mid",
            "mny"       => "application/x-msmoney",
            "mov"       => "video/quicktime",
            "movie"     => "video/x-sgi-movie",
            "mp2"       => "video/mpeg",
            "mp3"       => "audio/mpeg",
            "mpa"       => "video/mpeg",
            "mpe"       => "video/mpeg",
            "mpeg"      => "video/mpeg",
            "mpg"       => "video/mpeg",
            "mpp"       => "application/vnd.ms-project",
            "mpv2"      => "video/mpeg",
            "ms"        => "application/x-troff-ms",
            "mvb"       => "application/x-msmediaview",
            "nws"       => "message/rfc822",
            "oda"       => "application/oda",
            "p10"       => "application/pkcs10",
            "p12"       => "application/x-pkcs12",
            "p7b"       => "application/x-pkcs7-certificates",
            "p7c"       => "application/x-pkcs7-mime",
            "p7m"       => "application/x-pkcs7-mime",
            "p7r"       => "application/x-pkcs7-certreqresp",
            "p7s"       => "application/x-pkcs7-signature",
            "pbm"       => "image/x-portable-bitmap",
            "pdf"       => "application/pdf",
            "pfx"       => "application/x-pkcs12",
            "pgm"       => "image/x-portable-graymap",
            "pko"       => "application/ynd.ms-pkipko",
            "pma"       => "application/x-perfmon",
            "pmc"       => "application/x-perfmon",
            "pml"       => "application/x-perfmon",
            "pmr"       => "application/x-perfmon",
            "pmw"       => "application/x-perfmon",
            "pnm"       => "image/x-portable-anymap",
            "pot"       => "application/vnd.ms-powerpoint",
            "ppm"       => "image/x-portable-pixmap",
            "pps"       => "application/vnd.ms-powerpoint",
            "ppt"       => "application/vnd.ms-powerpoint",
            "prf"       => "application/pics-rules",
            "ps"        => "application/postscript",
            "pub"       => "application/x-mspublisher",
            "qt"        => "video/quicktime",
            "ra"        => "audio/x-pn-realaudio",
            "ram"       => "audio/x-pn-realaudio",
            "ras"       => "image/x-cmu-raster",
            "rgb"       => "image/x-rgb",
            "rmi"       => "audio/mid",
            "roff"      => "application/x-troff",
            "rtf"       => "application/rtf",
            "rtx"       => "text/richtext",
            "scd"       => "application/x-msschedule",
            "sct"       => "text/scriptlet",
            "setpay"    => "application/set-payment-initiation",
            "setreg"    => "application/set-registration-initiation",
            "sh"        => "application/x-sh",
            "shar"      => "application/x-shar",
            "sit"       => "application/x-stuffit",
            "snd"       => "audio/basic",
            "spc"       => "application/x-pkcs7-certificates",
            "spl"       => "application/futuresplash",
            "src"       => "application/x-wais-source",
            "sst"       => "application/vnd.ms-pkicertstore",
            "stl"       => "application/vnd.ms-pkistl",
            "stm"       => "text/html",
            "svg"       => "image/svg+xml",
            "sv4cpio"   => "application/x-sv4cpio",
            "sv4crc"    => "application/x-sv4crc",
            "t"         => "application/x-troff",
            "tar"       => "application/x-tar",
            "tcl"       => "application/x-tcl",
            "tex"       => "application/x-tex",
            "texi"      => "application/x-texinfo",
            "texinfo"   => "application/x-texinfo",
            "tgz"       => "application/x-compressed",
            "tif"       => "image/tiff",
            "tiff"      => "image/tiff",
            "tr"        => "application/x-troff",
            "trm"       => "application/x-msterminal",
            "tsv"       => "text/tab-separated-values",
            "txt"       => "text/plain",
            "uls"       => "text/iuls",
            "ustar"     => "application/x-ustar",
            "vcf"       => "text/x-vcard",
            "vrml"      => "x-world/x-vrml",
            "wav"       => "audio/x-wav",
            "wcm"       => "application/vnd.ms-works",
            "wdb"       => "application/vnd.ms-works",
            "wks"       => "application/vnd.ms-works",
            "wmf"       => "application/x-msmetafile",
            "wps"       => "application/vnd.ms-works",
            "wri"       => "application/x-mswrite",
            "wrl"       => "x-world/x-vrml",
            "wrz"       => "x-world/x-vrml",
            "xaf"       => "x-world/x-vrml",
            "xbm"       => "image/x-xbitmap",
            "xla"       => "application/vnd.ms-excel",
            "xlc"       => "application/vnd.ms-excel",
            "xlm"       => "application/vnd.ms-excel",
            "xls"       => "application/vnd.ms-excel",
            "xlsx"      => "vnd.ms-excel",
            "xlt"       => "application/vnd.ms-excel",
            "xlw"       => "application/vnd.ms-excel",
            "xof"       => "x-world/x-vrml",
            "xpm"       => "image/x-xpixmap",
            "xwd"       => "image/x-xwindowdump",
            "z"         => "application/x-compress",
            "zip"       => "application/zip",

            // Added
            "png"       => "image/png",
        );

        $exploded = explode('.', $file);
        $extension = end($exploded);
        $extension = strtolower($extension);
        return isset($mimeTypes[$extension]) ? $mimeTypes[$extension] : 'application/octet-stream';
    }

    protected function getKey(CRUDEntity $entity, $entityName, $field) {
        return $entity->getDefinition()->getFilePath($field).'/'.$entityName.'/'.$entity->get('id').'/'.$field;
    }

    public function __construct($bucket, $accessKey, $secretAccessKey) {
        $this->client = S3Client::factory(array(
            'key' => $accessKey,
            'secret' => $secretAccessKey
        ));
        $this->bucket = $bucket;
    }

    public function createFile(Request $request, CRUDEntity $entity, $entityName, $field) {
        $file = $request->files->get($field);
        if ($file) {
            $key = $this->getKey($entity, $entityName, $field).'/'.$file->getClientOriginalName();
            $result = $this->client->putObject(array(
                'Bucket' => $this->bucket,
                'Key' => $key,
                'SourceFile' => $file->getPathname()
            ));
        }
    }

    public function updateFile(Request $request, CRUDEntity $entity, $entityName, $field) {
        // We could first delete the old file, but for now, we are defensive and don't delete ever.
        $this->createFile($request, $entity, $entityName, $field);
    }

    public function deleteFile(CRUDEntity $entity, $entityName, $field) {
        // For now, we are defensive and don't delete ever.
    }

    public function renderFile(CRUDEntity $entity, $entityName, $field) {

        $fileName = $entity->get($field);
        $mimeType = $this->getMimeType($fileName);
        $key = $this->getKey($entity, $entityName, $field).'/'.$fileName;

        $result = $this->client->getObject(array(
            'Bucket' => $this->bucket,
            'Key'    => $key
        ));
        $result['Body']->rewind();

        $response = new StreamedResponse(function () use ($result) {
            while ($data = $result['Body']->read(1024)) {
                echo $data;
                flush();
            }
        }, 200, array(
            'Content-length' => $result['ContentLength'],
            'Content-Type' => $mimeType,
            'Content-Disposition' => 'attachment; filename="'.$fileName.'"'
        ));
        $response->send();

        return $response;
    }
}


src/CRUDlex/AmazonS3FileProcessor.php
<?php

/*
 * This file is part of the CRUDlex package.
 *
 * (c) Philip Lehmann-Böhm <philip@philiplb.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace CRUDlex;

use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpFoundation\StreamedResponse;
use Aws\S3\S3Client;

/**
 * An implementation of the FileProcessorInterface using the Amazon S3 service.
 */
class AmazonS3FileProcessor implements FileProcessorInterface {

    /**
     * The Amazon client having access to the bucket.
     */
    protected $client;

    /**
     * The bucket to store the files at.
     */
    protected $bucket;

    /**
     * Gets the S3 key of a file field.
     *
     * @param Entity $entity
     * the entity having the field
     * @param string $entityName
     * the name of the entity
     * @param string $field
     * the file field
     *
     * @return string
     * the key
     */
    protected function getKey(Entity $entity, $entityName, $field) {
        return $entity->getDefinition()->getField($field, 'path').'/'.$entityName.'/'.$entity->get('id').'/'.$field;
    }

    /**
     * Constructor.
     *
     * @param string $region
     * the region of the bucket
     * @param string $bucket
     * the bucket to store the files at
     * @param string $accessKey
     * the Amazon S3 Access Key
     * @param string $secretAccessKey
     * the Amazon S3 Secret Access Key
     */
    public function __construct($region, $bucket, $accessKey, $secretAccessKey) {
        $this->client = S3Client::factory([
            'region' => $region,
            'version' => '2006-03-01',
            'credentials' => [
                'key' => $accessKey,
                'secret' => $secretAccessKey
            ],
        ]);
        $this->bucket = $bucket;
    }

    /**
     * {@inheritdoc}
     */
    public function createFile(Request $request, Entity $entity, $entityName, $field) {
        $file = $request->files->get($field);
        if ($file) {
            $key = $this->getKey($entity, $entityName, $field).'/'.$file->getClientOriginalName();
            $result = $this->client->putObject([
                'Bucket' => $this->bucket,
                'Key' => $key,
                'SourceFile' => $file->getPathname()
            ]);
        }
    }

    /**
     * {@inheritdoc}
     */
    public function updateFile(Request $request, Entity $entity, $entityName, $field) {
        // We could first delete the old file, but for now, we are defensive and don't delete ever.
        $this->createFile($request, $entity, $entityName, $field);
    }

    /**
     * {@inheritdoc}
     */
    public function deleteFile(Entity $entity, $entityName, $field) {
        // For now, we are defensive and don't delete ever.
    }

    /**
     * {@inheritdoc}
     */
    public function renderFile(Entity $entity, $entityName, $field) {

        $fileName = $entity->get($field);
        $mimeTypes = new MimeTypes();
        $mimeType = $mimeTypes->getMimeTypeByExtension($fileName);
        $key = $this->getKey($entity, $entityName, $field).'/'.$fileName;

        $result = $this->client->getObject([
            'Bucket' => $this->bucket,
            'Key'    => $key
        ]);
        $result['Body']->rewind();

        $response = new StreamedResponse(function () use ($result) {
            while ($data = $result['Body']->read(1024)) {
                echo $data;
                flush();
            }
        }, 200, [
            'Cache-Control' => 'public, max-age=86400',
            'Content-length' => $result['ContentLength'],
            'Content-Type' => $mimeType,
            'Content-Disposition' => 'attachment; filename="'.$fileName.'"'
        ]);
        $response->send();

        return $response;
    }
}
